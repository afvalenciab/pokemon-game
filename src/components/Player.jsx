import React, { useRef, useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { requestMorePokemons, getPokemonsError } from '../actions';
import Carousel from '../components/Carousel';
import CarouselItem from '../components/CarouselItem';
import PokemonCard from '../components/PokemonCard';
import logoPokemon from '../assets/static/Pokemon_logo.svg';
import '../assets/styles/Player.scss';

const Player = props => {
  const {
    pokemonList,
    requestMorePokemons,
    nextUrlPokemonsList,
    getPokemonsError
  } = props;

  const loaderUrl = useRef(nextUrlPokemonsList);
  const [element, setElement] = useState(null);

  const observer = useRef(
    new IntersectionObserver(
      entries => {
        if (entries[0].isIntersecting) {
          requestMorePokemons(loaderUrl.current);
        }
      },
      { rootMargin: '0px 500px 0px 0px' }
    )
  );

  useEffect(() => {
    const currentElement = element;
    const currentObserver = observer.current;

    if (currentElement) {
      currentObserver.observe(currentElement);
    }

    return () => {
      if (currentElement) {
        currentObserver.unobserve(currentElement);
      }
    };
  }, [element]);

  useEffect(() => {
    loaderUrl.current = nextUrlPokemonsList;
  }, [nextUrlPokemonsList]);

  const [searchPokemon, setSearchPokemon] = useState({
    isSearching: false,
    pokemonSearched: {},
    nameOrId: '',
    message: ''
  });

  const handleInput = (event) => {
    setSearchPokemon({
      ...searchPokemon,
      nameOrId: event.target.value
    });
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    if (searchPokemon.nameOrId) {
      setSearchPokemon({
        ...searchPokemon,
        isSearching: true,
        message: 'Loading...'
      });

      const fetchData = async () => {
        try {
          const nameOrIdPokemon = searchPokemon.nameOrId.toLowerCase();
          const response = await fetch(
            `https://pokeapi.co/api/v2/pokemon/${nameOrIdPokemon}`
          );
          const data = await response.json();
          setSearchPokemon({
            ...searchPokemon,
            isSearching: true,
            pokemonSearched: data
          });
        } catch (error) {
          getPokemonsError(error);

          setSearchPokemon({
            ...searchPokemon,
            isSearching: true,
            pokemonSearched: {},
            message: 'No encontrado'
          });
        }
      };

      fetchData();
    } else {
      setSearchPokemon({
        ...searchPokemon,
        isSearching: false
      });
    }
  };

  return (
    <section className='home--player'>
      <h2>Selecciona un Pokémon</h2>
      {pokemonList.length > 0 && (
        <Carousel>
          {!searchPokemon.isSearching &&
            pokemonList.map((item, index) => (
              <CarouselItem key={index} {...item} />
            ))}

          {!searchPokemon.isSearching && (
            <div className='observer' ref={setElement}></div>
          )}

          {searchPokemon.isSearching &&
            (Object.keys(searchPokemon.pokemonSearched).length > 0 ? (
              <PokemonCard
                name={searchPokemon.pokemonSearched.name}
                urlPicture={searchPokemon.pokemonSearched.sprites.front_default}
              />
            ) : (
              <PokemonCard name={searchPokemon.message} urlPicture={logoPokemon} />
            ))}
        </Carousel>
      )}
      <div className='player--search'>
        <form onSubmit={handleSubmit}>
          <input
            type='text'
            placeholder='Buscar Pokémon por nombre o Id'
            onChange={handleInput}
          />
          <button type='submit'>⟶</button>
        </form>
      </div>
    </section>
  );
};

const mapStateToProps = state => {
  return {
    pokemonList: state.pokemonList,
    nextUrlPokemonsList: state.nextUrlPokemonsList
  };
};

const mapDispatchToProps = {
  requestMorePokemons,
  getPokemonsError
};

export default connect(mapStateToProps, mapDispatchToProps)(Player);
