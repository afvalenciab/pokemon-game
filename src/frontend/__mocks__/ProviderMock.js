import React from 'react';
import { createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { Router } from 'react-router';
import { Provider } from 'react-redux';
import { createBrowserHistory } from 'history';
import initialState from './initialState';
import reducer from '../reducers';

const store = createStore(reducer, initialState, compose(applyMiddleware(thunk)));
const history = createBrowserHistory();

const ProviderMock = (props) => {
  return(
    <Provider store={store}>
      <Router history={history}>
        {props.children}
      </Router>
    </Provider>
  );
};

export default ProviderMock;