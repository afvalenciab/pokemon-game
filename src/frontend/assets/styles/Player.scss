
.home--player {
  background: #FFF;
  border-top-left-radius: 30px;
  border-bottom-right-radius: 30px;
  display: grid;
  grid-template-rows: 45px 1fr 45px;
  grid-gap: 5px;
  height: 100%;
  padding: 10px;
  justify-items: center;
  align-items: center;
  box-sizing: border-box;
  box-shadow: 0px 0px 7px -1px rgba(0, 0, 0, 0.5);

  h2 {
    margin: 0;
  }

  .player--search {
    form {
      display: flex;
  
      input{
        outline: none;
        width: 240px;
        height: 30px;
        align-self: flex-start;
        border: 0;
        border: 1px solid gray;
        border-right: 0;
        border-radius: 25px 0 0 25px;
        padding-left: 10px;
        font-size: 15px;
      }
  
      button {
        border: 0;
        border: 1px solid gray;
        border-left: 0;
        background-color: #3C5BA6;
        color: #FFCB07;
        font-size: 15px;
        font-weight: bold;
        height: 34px;
        width: 40px;
        border-radius: 0 25px 25px 0;
        cursor: pointer;
        outline: none;
        transition: .2s;
        padding: 0;
      }
      button:hover {
        background-color: rgb(50, 76, 138);
      }
      button:active {
        font-size: 13px;
      }
    }
  }
}
