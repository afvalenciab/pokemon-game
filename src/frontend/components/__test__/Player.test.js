import React from 'react';
import { create } from 'react-test-renderer';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ProviderMock from '../../__mocks__/ProviderMock';
import Player from '../Player';

configure({ adapter: new Adapter() });

describe('Player testing', () => {
  test('Match snaptshot', () => {
    const player = create(
      <ProviderMock>
        <Player />
      </ProviderMock>
    );
    expect(player.toJSON()).toMatchSnapshot();
  });

  test('Calls and execute preventDefault function onSubmit form', () => {
    const preventDefault = jest.fn();
    const player = mount(
      <ProviderMock>
        <Player />
      </ProviderMock>
    );
    player.find('form').simulate('submit', { preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    player.unmount();
  });
});