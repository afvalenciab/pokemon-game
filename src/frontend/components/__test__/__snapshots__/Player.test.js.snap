// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player testing Match snaptshot 1`] = `
<section
  className="home--player"
>
  <h2>
    Selecciona un Pokémon
  </h2>
  <div
    className="player--search"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Buscar Pokémon por nombre o Id"
        type="text"
      />
      <button
        type="submit"
      >
        ⟶
      </button>
    </form>
  </div>
</section>
`;
